### ISSUE TITLE
Automatically detect Linux install type

### ISSUE THREAD

**Issue by scudette** [Velocidex/velociraptor#3451]
Currently there are two flavors of Linux startup services: Sysv is for older mostly redhat systems and systemd is for more modern systems. We can build two types of RPM using the --sysv flag.

However in practice it is sometimes difficult to selectively send the right type of RPM to the right systems - it would be easier if we could just send the same RPM to all systems and have it figure out which type of startup script to use based on the system itself.

Maybe this should be done by presubmit scripts. This will make is a lot easier to deploy as we can use the same RPM to all systems in the fleet.

**Comment by freeload101** [comment]
Yah guessign it would be a pre-install script you add to the RPM then just have it push the right conf by adding whatever is missing from the orig RPM? 

reminds me of the Splunk UF script I made :
https://github.com/freeload101/SCRIPTS/blob/master/Bash/Splunk_UniversalForwarder_Installer.bash

**Comment by misje** [comment]
I would just like to show what debhelper does (using both dh_installinit and dh_installsystemd). Normally when building a deb package, one would let debhelper do all the heavy packaging work. debhelper tries to do the right thing in a policy-compliant way. 

Let's say my package (and by default, also my initv script / service file) name is foobar. The following is added to the end of preinst:

```bash
# Automatically added by dh_installinit/13.6ubuntu1
if [ -z "${DPKG_ROOT:-}" ] && [ "$1" = upgrade ] && [ -d /run/systemd/system ] ; then
	deb-systemd-invoke stop 'foobar.service' >/dev/null || true
fi
# End automatically added section
```

Added to postinst:

```bash
# Automatically added by dh_installinit/13.6ubuntu1
if [ "$1" = "configure" ] || [ "$1" = "abort-upgrade" ] || [ "$1" = "abort-deconfigure" ] || [ "$1" = "abort-remove" ] ; then
	if [ -x "/etc/init.d/foobar" ]; then
		update-rc.d foobar defaults-disabled >/dev/null || exit 1
	fi
fi
# End automatically added section
# Automatically added by dh_installsystemd/13.6ubuntu1
if [ "$1" = "configure" ] || [ "$1" = "abort-upgrade" ] || [ "$1" = "abort-deconfigure" ] || [ "$1" = "abort-remove" ] ; then
	if deb-systemd-helper debian-installed 'foobar.service'; then
		# This will only remove masks created by d-s-h on package removal.
		deb-systemd-helper unmask 'foobar.service' >/dev/null || true

		if deb-systemd-helper --quiet was-enabled 'foobar.service'; then
			# Create new symlinks, if any.
			deb-systemd-helper enable 'foobar.service' >/dev/null || true
		fi
	fi

	# Update the statefile to add new symlinks (if any), which need to be cleaned
	# up on purge. Also remove old symlinks.
	deb-systemd-helper update-state 'foobar.service' >/dev/null || true
fi
# End automatically added section
```

Added to postrm:

```bash
# Automatically added by dh_installinit/13.6ubuntu1
if [ -z "${DPKG_ROOT:-}" ] && [ "$1" = "purge" ] ; then
	update-rc.d foobar remove >/dev/null
fi
# End automatically added section
# Automatically added by dh_installsystemd/13.6ubuntu1
if [ "$1" = remove ] && [ -d /run/systemd/system ] ; then
	systemctl --system daemon-reload >/dev/null || true
fi
# End automatically added section
# Automatically added by dh_installsystemd/13.6ubuntu1
if [ "$1" = "remove" ]; then
	if [ -x "/usr/bin/deb-systemd-helper" ]; then
		deb-systemd-helper mask 'foobar.service' >/dev/null || true
	fi
fi

if [ "$1" = "purge" ]; then
	if [ -x "/usr/bin/deb-systemd-helper" ]; then
		deb-systemd-helper purge 'foobar.service' >/dev/null || true
		deb-systemd-helper unmask 'foobar.service' >/dev/null || true
	fi
fi
# End automatically added section
```

**Comment by scudette** [comment]
It looks like deb-systemd-helper is only available on debian systems right? This issue is mainly around rpm since we dont offer sysv scripts for debian systems by default. Perhaps we should though?

**Comment by misje** [comment]
I don't think you truly need that helper command. It just adds some safeguards, and it is a fairly small bash script. Some notes from the man page:

- The "enable" action will only be performed once (when first installing the package). On the first "enable", a state file is created which will be deleted upon "purge"
- The "mask" action will keep state on whether the service was enabled/disabled before and will properly return to that state on "unmask".

I can't think of issues with running `systemctl enable` (or any of the other commands) twice.

I just added the contents from debhelper as a reference for manual re-implementation, in order to not miss any important commands. I think you'll have to go a number of versions back in order to find a non-systemd Debian system, but I'm sure there are users out there running velociraptor on truly ancient systems that could benefit of sysv support.

**Comment by misje** [comment]
My bad, there are some differences that I skipped. Like this one, which is used in the script snippet above:

> The "was-enabled" action is not present in systemctl, but is required in Debian so that we can figure out whether a service was enabled before we installed an updated service file. See http://bugs.debian.org/717603 for details.

**Comment by scudette** [comment]
This is now released in 0.73-rc1
